src/mongo/db/pipeline/expression.cpp

```diff
@@ -3072,6 +3085,123 @@ const char* ExpressionWeek::getOpName() const {
     return "$week";
 }
 
+/* ------------------------- ExpressionIsoDayOfWeek --------------------- */
+
+Value ExpressionIsoDayOfWeek::evaluateInternal(Variables* vars) const {
+    Value pDate(vpOperand[0]->evaluateInternal(vars));
+    return Value(extract(pDate.coerceToTm()));
+}
+
+int ExpressionIsoDayOfWeek::extract(const tm& tm) {
+    return (tm.tm_wday - 7) % 7 + 7;
+}
+
+REGISTER_EXPRESSION(isoDayOfWeek, ExpressionIsoDayOfWeek::parse);
+const char* ExpressionIsoDayOfWeek::getOpName() const {
+    return "$isoDayOfWeek";
+}
+
+/* ------------------------- ExpressionIsoWeekYear ---------------------- */
+
+Value ExpressionIsoWeekYear::evaluateInternal(Variables* vars) const {
+    Value pDate(vpOperand[0]->evaluateInternal(vars));
+    return Value(extract(pDate.coerceToTm()));
+}
+
+int ExpressionIsoWeekYear::extract(const tm& tm) {
+    if (tm.tm_mon > 0 && tm.tm_mon < 11) {
+        // if month isnt January or December it's just the year given
+        return tm.tm_year + 1900;
+    } else if (tm.tm_mon == 0) {
+        // In January we  need to check if the week belongs to previous year
+        int isoWeek = ExpressionIsoWeek::extract(tm);
+        if (isoWeek > 51) { // weeks 52 and 53 belong to the previous year
+            return tm.tm_year + 1900 - 1;
+        } else { // all other weeks belong to given year
+            return tm.tm_year + 1900;
+        }
+    } else {
+        // the week 1 in December belongs to the next year
+        int isoWeek = ExpressionIsoWeek::extract(tm);
+        if (isoWeek == 1) {
+            return tm.tm_year + 1900 + 1;
+        } else {
+            return tm.tm_year + 1900;
+        }
+    }
+}
+
+REGISTER_EXPRESSION(isoWeekYear, ExpressionIsoWeekYear::parse);
+const char* ExpressionIsoWeekYear::getOpName() const {
+    return "$isoWeekYear";
+}
+
+/* ------------------------- ExpressionIsoWeek -------------------------- */
+
+Value ExpressionIsoWeek::evaluateInternal(Variables* vars) const {
+    Value pDate(vpOperand[0]->evaluateInternal(vars));
+    return Value(extract(pDate.coerceToTm()));
+}
+
+int ExpressionIsoWeek::lastWeek(int year) {
+    // create YYYY-12-31T23:59:59 so only 1 second left to new year
+    struct tm tm;
+    tm.tm_year = year - 1900;
+    tm.tm_mon = 11;
+    tm.tm_mday = 31;
+    tm.tm_hour = 23;
+    tm.tm_min = 59;
+    tm.tm_sec = 59;
+    mktime(&tm);
+    /* From: https://en.wikipedia.org/wiki/ISO_week_date#Last_week
```

Comment:
Please use `// Comment` style comments when the comment is not documenting a method, and
```
/**
 * Comment
 */
```
style comments when the comment is documenting a method.